src/protocols/multi_party_ecdsa/gg_2020/test.rs

```diff
@@ -150,6 +191,210 @@ fn test_sign_n5_t2_ttag4_corrupt_step7_party24() {
     assert!(&res.err().unwrap().bad_actors[..] == &[1, 3])
 }
 
+//
+// As per page13 https://eprint.iacr.org/2020/540.pdf:
+// This step will:
+// 1. This participant will create a Commitment, Decommitment pair on a scalar
+//    ui and then publish the Commitment part.
+// 2. It will create a Paillier Keypair and publish the public key for that.
+//
+#[cfg(test)]
+fn keygen_stage1(
+    participant: usize,
+) -> (
+    Keys,
+    KeyGenBroadcastMessage1,
+    KeyGenDecommitMessage1,
+    DLogStatement,
+) {
+    // Paillier keys and various other values
+    // party_keys.ek is a secret value and it should be encrypted
+    // using a key that is owned by the participant who creates it. Right now it's plaintext but
+    // this is test.
+    //
+    let party_keys = Keys::create(participant - 1);
+    let (bc1, decom) =
+        party_keys.phase1_broadcast_phase3_proof_of_correct_key_proof_of_correct_h1h2();
+    let h1_h2_N_tilde = bc1.dlog_statement.clone();
+    (party_keys, bc1, decom, h1_h2_N_tilde)
+}
+
+//
+// As per page 13 on https://eprint.iacr.org/2020/540.pdf:
+// 1. Decommit the value obtained in stage1.
+// 2. Perform a VSS on that value.
+//
+#[cfg(test)]
+fn keygen_stage2(
+    participant: usize,
+    params: &Parameters,
+    party_keys: &[Keys],
+    bc1_vec: &[KeyGenBroadcastMessage1],
+    decom_vec: &[KeyGenDecommitMessage1],
+) -> (VerifiableSS, Vec<FE>, usize) {
+    let vss_result = party_keys[participant - 1]
+        .phase1_verify_com_phase3_verify_correct_key_verify_dlog_phase2_distribute(
+            params, decom_vec, bc1_vec,
+        )
+        .expect("stage 2 keygen failed");
+    let (vss_scheme, secret_shares, index) = vss_result;
+    (vss_scheme, secret_shares, index)
+}
+
+//
+// As per page 13 on https://eprint.iacr.org/2020/540.pdf:
+// 1. Participant adds there private shares to obtain their final share of the keypair.
+// 2. Calculate the corresponding public key for that share.
+// 3. Generate the dlog proof which the orchestrator would check later.
+//
+#[cfg(test)]
+fn keygen_stage3(
+    party_keys: &Keys,
+    vss_scheme_vec: &[VerifiableSS],
+    secret_shares_vec: &Vec<Vec<FE>>,
+    decom_vec: &[KeyGenDecommitMessage1],
+    params: &Parameters,
+    participant: usize,
+    index_vec: &[usize],
+) -> (SharedKeys, DLogProof) {
+    let y_vec = (0..params.share_count)
+        .map(|i| decom_vec[i as usize].y_i)
+        .collect::<Vec<GE>>();
+    let res = party_keys
+        .phase2_verify_vss_construct_keypair_phase3_pok_dlog(
+            &params,
+            &y_vec,
+            &secret_shares_vec[participant - 1],
+            vss_scheme_vec,
+            &index_vec[participant - 1] + 1,
+        )
+        .expect("stage3 failed.");
+    let (shared_keys, dlog_proof) = res;
+    (shared_keys, dlog_proof)
+}
+
+//
+// The key generation process needs to be orchestrated.
```

Comment:
orchestration is optional. It can be done in a p2p fashion as well.

Reply:
right. Updating comment.